src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,370 @@
+    try (final ZipFile zipFile = new ZipFile(pulledFile)) {
+      final Enumeration<? extends ZipEntry> enumeration = zipFile.entries();
+      while (enumeration.hasMoreElements()) {
+        final ZipEntry entry = enumeration.nextElement();
+        size += FileUtils.retryCopy(
+            new StreamUtils.InputStreamFactory<InputStream>()
+            {
+              @Override
+              public InputStream build()
+              {
+                try {
+                  return new BufferedInputStream(zipFile.getInputStream(entry));
+                }
+                catch (IOException e) {
+                  throw new IAE(e, "Bad entry: [%s]", entry.getName());
+                }
+              }
+            },
+            FileUtils.MD5ERROR,
+            Paths.get(outDir.getAbsolutePath(), entry.getName()).toFile(),
+            10
+        );
+      }
+    }
+    return size;
+  }
+
+  public static long unzip(InputStream in, File outDir) throws IOException
+  {
+    long size = 0l;
+    try (final ZipInputStream zipIn = new ZipInputStream(in)) {
+      ZipEntry entry;
+      while ((entry = zipIn.getNextEntry()) != null) {
+        try (FileOutputStream out = new FileOutputStream(new File(outDir, entry.getName()))) {
+          size += ByteStreams.copy(zipIn, out);
+          zipIn.closeEntry();
+        }
+      }
+    }
+    return size;
+  }
+
+  public static long gunzip(final File pulledFile, File outDir) throws IOException
+  {
+    final String name = pulledFile.getName();
+    if (!name.endsWith(".gz")) {
+      throw new IAE("[%s] is not a .gz file", pulledFile.getAbsolutePath());
+    }
+    final File outFile = Paths.get(outDir.getAbsolutePath(), name.substring(0, name.length() - ".gz".length()))
+                              .toFile();
+    log.info("Gunzipping file[%s] to [%s]", pulledFile.getAbsolutePath(), outFile.getAbsolutePath());
+    return gunzip(
+        new StreamUtils.InputStreamFactory<InputStream>()
+        {
+          @Override
+          public InputStream build() throws IOException
+          {
+            try {
+              return new FileInputStream(pulledFile);
+            }
+            catch (FileNotFoundException e) {
+              throw new IAE(e, "Could not find file [%s]", pulledFile.getAbsolutePath());
+            }
+          }
+        },
+        outFile
+    );
+  }
+
+  /**
+   * Unzips the input stream via a gzip filter
+   *
+   * @param in      The input stream to run through the gunzip filter. This stream is closed
+   * @param outFile The file to output to
+   *
+   * @throws IOException
+   */
+  public static long gunzip(InputStream in, File outFile) throws IOException
+  {
+    try (GZIPInputStream gzipInputStream = new GZIPInputStream(fix7036144(in))) {
+      final ByteSink sink = Files.asByteSink(outFile);
+      return sink.writeFrom(gzipInputStream);
+    }
+  }
+
+  public static InputStream fix7036144(final InputStream in)
```

Comment:
Bug is referenced below, but I'll put it in the method comments and change the name.